server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1539,6 +1557,126 @@ public RepositoriesService getRepositoriesService() {
+                        }
+                        shardsBuilder.put(shardId, newStatus);
+                    }
+                }
+            }
+            if (shardsBuilder == null) {
+                entries.add(entry);
+            } else {
+                snapshotsInProgressChanged = true;
+                ImmutableOpenMap<ShardId, ShardSnapshotStatus> shards = shardsBuilder.build();
+                entries.add(
+                    new SnapshotsInProgress.Entry(
+                        entry,
+                        completed(shards.values()) ? State.SUCCESS : entry.state(),
+                        shards
+                    )
+                );
+            }
+        }
+        if (snapshotsInProgressChanged) {
+            return new SnapshotsInProgress(entries);
+        }
+        return oldSnapshot;
+    }
+
+    private static ShardSnapshotStatus determineShardSnapshotStatus(final ShardSnapshotStatus currentStatus,
+        final ShardRouting primaryShardRouting) {
+        final State currentState = currentStatus.state();
+        final ShardSnapshotStatus newStatus;
+        if (primaryShardRouting == null) {
+            newStatus = failedStatus(null, "missing shard");
+        } else if (primaryShardRouting.active() == false) {
+            if (primaryShardRouting.initializing() && currentState == State.WAITING) {
+                newStatus = currentStatus;
+            } else {
+                newStatus = failedStatus(
+                    primaryShardRouting.currentNodeId(),
+                    primaryShardRouting.unassignedInfo().getReason().toString()
+                );
+            }
+        } else if (primaryShardRouting.started()) {
+            switch (currentState) {
+                case WAITING:
+                    newStatus = new ShardSnapshotStatus(primaryShardRouting.currentNodeId());
+                    break;
+                case INIT: {
+                    String currentNodeId = currentStatus.nodeId();
+                    assert currentNodeId != null;
+                    if (primaryShardRouting.currentNodeId().equals(currentNodeId)) {
+                        newStatus = currentStatus;
+                    } else {
+                        newStatus = failedStatus(currentNodeId);
+                    }
+                    break;
+                }
+                case ABORTED:
+                    String currentNodeId = currentStatus.nodeId();
+                    if (currentNodeId.equals(primaryShardRouting.currentNodeId())) {
+                        newStatus = currentStatus;
+                    } else {
+                        newStatus = failedStatus(currentNodeId);
+                    }
+                    break;
+                default:
+                    newStatus = currentStatus;
+                    break;
+            }
+        } else if (currentState == State.INIT || currentStatus.state() == State.ABORTED) {
```

Comment:
can you add a comment that we are in the relocating state here

Reply:
Added the assertion for that below.

Reply:
perhaps move the assertion one level up, i.e.,

```
} else {
  assert priamryShardRouting.relocating();
  if (currentState == State.INIT || currentStatus.state() == State.ABORTED) { 
    ..
  } else {
    ..
  }
}
```